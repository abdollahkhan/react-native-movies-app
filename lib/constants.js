export const API_URL = 'https://api.themoviedb.org/3'
export const API_KEY = '0511c9bf8a6b1d1216961e19831d9a40'

export const SCREENS = {
  SETTINGS: 'Settings',
  FAVOURITES: 'Favourites',
  DISCOVER: 'Discover',
  MAIN: 'Main'
}
