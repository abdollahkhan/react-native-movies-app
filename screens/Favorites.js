import Icon from 'react-native-vector-icons/Ionicons'
import {
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native'
import { FeaturedImage } from '../components'
import EmptyState from '../components/EmptyState'
import { colors, sizes, textStyles } from '../lib/styles'
import { useDispatch, useSelector } from 'react-redux'
import { selectFavourites } from '../redux/selectors/favaourite.selector'
import { removeFavourite } from '../redux/reducers/favourite.reducer'
import {
  selectMediaBaseUrl,
  selectMediaSizes
} from '../redux/selectors/media.selector'
import { SCREENS } from '../lib/constants'

export default function Favorites({ navigation }) {
  const favourites = useSelector(selectFavourites)
  const dispatch = useDispatch()

  const mediaUrl = useSelector(selectMediaBaseUrl)
  const mediaSizes = useSelector(selectMediaSizes)

  const isEmpty = favourites.length === 0

  const removeFav = (id) => {
    dispatch(removeFavourite(id))
  }

  return (
    <ScrollView
    style={{backgroundColor: colors.white}}
      contentContainerStyle={[isEmpty && styles.empty, styles.container]}
    >
      {(isEmpty && (
        <EmptyState
          image={require('../assets/empty-favorites.jpg')}
          title="You haven't liked any movie yet"
          message="Why not try to find a movie you like?"
          actionLabel="Go to Discover"
          onAction={() => navigation.navigate(SCREENS.DISCOVER)}
        />
      )) ||
        favourites?.map((fav) => {
          return (
            <View style={styles.column} key={fav.id}>
              <TouchableOpacity
                style={styles.favIconContainer}
                onPress={() => removeFav(fav.id)}
              >
                <Icon
                  name={'heart'}
                  size={sizes.xxl}
                  color={colors.primary}
                  style={styles.favIcon}
                />
              </TouchableOpacity>
              <FeaturedImage
                src={`${mediaUrl}${mediaSizes?.[2]}${fav.posterPath}`}
              />
              <Text style={textStyles.h3}>{fav.title}</Text>
            </View>
          )
        })}
    </ScrollView>
  )
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    flexDirection: 'row',
    flexWrap: 'wrap',
    height: '100%',
    padding: 10,
    justifyContent: 'space-between',
    backgroundColor: colors.white
  },

  empty: {
    justifyContent: 'center'
  },

  column: {
    width: '48%',
    position: 'relative'
  },

  content: {
    width: '70%'
  },

  favIcon: {
    position: 'absolute',
    top: 2,
    bottom: 0,
    left: 3,
    right: 0
  },

  favIconContainer: {
    position: 'absolute',
    right: 5,
    top: 5,
    padding: 15,
    backgroundColor: colors.white,
    zIndex: 2,
    borderRadius: 25
  }
})
