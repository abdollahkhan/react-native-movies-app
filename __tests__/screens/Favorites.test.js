import React from 'react'
import Favorites from '../../screens/Favorites'
import { configureStore } from '@reduxjs/toolkit'
import favouriteReducer from '../../redux/reducers/favourite.reducer'
import mediaReducer from '../../redux/reducers/media.reducer'
import { Provider } from 'react-redux'
import {
  render,
  screen,
  fireEvent,
  cleanup
} from '@testing-library/react-native'
import { SCREENS } from '../../lib/constants'

describe('Favorites Screen', () => {
  const getStore = (initialState) => {
    const store = configureStore({
      reducer: { favourites: favouriteReducer, media: mediaReducer },
      preloadedState: initialState
    })

    return store
  }

  afterEach(() => {
    jest.restoreAllMocks()
    cleanup()
  })

  const mockNavigation = {
    navigate: jest.fn()
  }

  it('renders correctly when favourites is empty', async () => {
    const store = getStore({
      favourites: {
        data: {}
      }
    })

    render(
      <Provider store={store}>
        <Favorites navigation={mockNavigation} />
      </Provider>
    )

    expect(
      await screen.findByText('Why not try to find a movie you like?')
    ).toBeDefined()
    expect(screen).toMatchSnapshot()
  })

  it('should navigate to discover screen when favourites is empty', async () => {
    const store = getStore({
      favourites: {
        data: {}
      }
    })

    render(
      <Provider store={store}>
        <Favorites navigation={mockNavigation} />
      </Provider>
    )

    fireEvent(screen.getByText('Go to Discover'), 'click')
    expect(mockNavigation.navigate).toBeCalledWith(SCREENS.DISCOVER)
  })

  it('renders correctly when favourites is not empty', async () => {
    const store = getStore({
      favourites: {
        data: {
          1: {
            id: 1,
            title: 'Mock Movie',
            posterPath: '/posterPath1.jpg'
          },
          2: {
            id: 2,
            title: 'Mock Movie',
            posterPath: '/posterPath2.jpg'
          }
        }
      }
    })

    render(
      <Provider store={store}>
        <Favorites navigation={mockNavigation} />
      </Provider>
    )

    expect(await screen.findAllByText('Mock Movie')).toHaveLength(2)
    expect(screen).toMatchSnapshot()
  })

  it('should remove a item from favourites', async () => {
    const store = getStore({
      favourites: {
        data: {
          1: {
            id: 1,
            title: 'Mock Movie',
            posterPath: '/posterPath1.jpg'
          },
          2: {
            id: 2,
            title: 'Mock Movie',
            posterPath: '/posterPath2.jpg'
          }
        }
      }
    })

    render(
      <Provider store={store}>
        <Favorites navigation={mockNavigation} />
      </Provider>
    )

    expect(await screen.findAllByText('Mock Movie')).toHaveLength(2)
    fireEvent(screen.getAllByLabelText('Remove Favorite Button')[0], 'click')

    const favs = Object.values(store.getState().favourites.data)
    expect(favs).toHaveLength(1)
  })
})
