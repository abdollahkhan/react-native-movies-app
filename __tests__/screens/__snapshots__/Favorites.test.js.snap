// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites Screen renders correctly when favourites is empty 1`] = `
<RCTScrollView
  contentContainerStyle={
    [
      {
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#ffffff",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": "100%",
        "justifyContent": "space-between",
        "padding": 10,
        "width": "100%",
      },
    ]
  }
  style={
    {
      "backgroundColor": "#ffffff",
    }
  }
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "gap": 10,
          "justifyContent": "center",
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../assets/empty-favorites.jpg",
          }
        }
        style={
          {
            "height": 150,
            "marginBottom": 10,
            "width": 200,
          }
        }
      />
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 18,
            "fontWeight": "700",
            "marginBottom": 0,
          }
        }
      >
        You haven't liked any movie yet
      </Text>
      <Text
        style={
          {
            "color": "#757575",
            "fontSize": 14,
            "marginBottom": 14,
            "marginTop": 0,
          }
        }
      >
        Why not try to find a movie you like?
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#000000",
            "borderRadius": 16,
            "opacity": 1,
            "paddingHorizontal": 30,
            "paddingVertical": 16,
          }
        }
      >
        <Text
          style={
            {
              "color": "#ffffff",
              "fontSize": 14,
              "fontWeight": "700",
            }
          }
        >
          Go to Discover
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Favorites Screen renders correctly when favourites is not empty 1`] = `
<RCTScrollView
  contentContainerStyle={
    [
      false,
      {
        "backgroundColor": "#ffffff",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": "100%",
        "justifyContent": "space-between",
        "padding": 10,
        "width": "100%",
      },
    ]
  }
  style={
    {
      "backgroundColor": "#ffffff",
    }
  }
>
  <View>
    <View
      style={
        {
          "position": "relative",
          "width": "48%",
        }
      }
    >
      <View
        accessibilityLabel="Remove Favorite Button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 25,
            "opacity": 1,
            "padding": 15,
            "position": "absolute",
            "right": 5,
            "top": 5,
            "zIndex": 2,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#0d9488",
                "fontSize": 24,
              },
              {
                "bottom": 0,
                "left": 3,
                "position": "absolute",
                "right": 0,
                "top": 2,
              },
              {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          ïŒ®
        </Text>
      </View>
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          {
            "aspectRatio": 0.7142857142857143,
            "backgroundColor": "transparent",
            "borderRadius": 12,
            "overflow": "hidden",
            "position": "relative",
            "width": "100%",
          }
        }
      >
        <Image
          onLoad={[Function]}
          source={
            {
              "uri": "undefinedundefined/posterPath1.jpg",
            }
          }
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          testID="RNE__Image"
          theme={
            {
              "colors": {
                "background": "#ffffff",
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": {
                  "android": {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ad1457",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
              "spacing": {
                "lg": 12,
                "md": 8,
                "sm": 4,
                "xl": 24,
                "xs": 2,
              },
            }
          }
          transitionDuration={360}
        />
        <View
          accessibilityElementsHidden={true}
          collapsable={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#bdbdbd",
                "justifyContent": "center",
              }
            }
            testID="RNE__Image__placeholder"
          >
            <View
              accessibilityLabel="loading..."
              accessibilityRole="none"
              accessible={false}
              onLayout={[Function]}
              style={
                [
                  {
                    "borderRadius": 2,
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#bdc6cf",
                    "height": 12,
                    "width": "100%",
                  },
                  undefined,
                  {
                    "height": "100%",
                    "width": "100%",
                  },
                ]
              }
              testID="RNE__Skeleton"
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgb(166, 178, 190)",
                    "height": "100%",
                    "transform": [
                      {
                        "translateX": -0,
                      },
                    ],
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={{}}
          testID="RNE__Image__children__container"
        />
      </View>
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 16,
            "fontWeight": "700",
            "marginBottom": 8,
          }
        }
      >
        Mock Movie
      </Text>
    </View>
    <View
      style={
        {
          "position": "relative",
          "width": "48%",
        }
      }
    >
      <View
        accessibilityLabel="Remove Favorite Button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 25,
            "opacity": 1,
            "padding": 15,
            "position": "absolute",
            "right": 5,
            "top": 5,
            "zIndex": 2,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            [
              {
                "color": "#0d9488",
                "fontSize": 24,
              },
              {
                "bottom": 0,
                "left": 3,
                "position": "absolute",
                "right": 0,
                "top": 2,
              },
              {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          ïŒ®
        </Text>
      </View>
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          {
            "aspectRatio": 0.7142857142857143,
            "backgroundColor": "transparent",
            "borderRadius": 12,
            "overflow": "hidden",
            "position": "relative",
            "width": "100%",
          }
        }
      >
        <Image
          onLoad={[Function]}
          source={
            {
              "uri": "undefinedundefined/posterPath2.jpg",
            }
          }
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          testID="RNE__Image"
          theme={
            {
              "colors": {
                "background": "#ffffff",
                "black": "#242424",
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": {
                  "android": {
                    "error": "#f44336",
                    "grey": "rgba(0, 0, 0, 0.54)",
                    "primary": "#2196f3",
                    "searchBg": "#dcdce1",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "default": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "ios": {
                    "error": "#ff3b30",
                    "grey": "#7d7d7d",
                    "primary": "#007aff",
                    "searchBg": "#dcdce1",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                  "web": {
                    "error": "#ff190c",
                    "grey": "#393e42",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ad1457",
                "success": "#52c41a",
                "warning": "#faad14",
                "white": "#ffffff",
              },
              "spacing": {
                "lg": 12,
                "md": 8,
                "sm": 4,
                "xl": 24,
                "xs": 2,
              },
            }
          }
          transitionDuration={360}
        />
        <View
          accessibilityElementsHidden={true}
          collapsable={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#bdbdbd",
                "justifyContent": "center",
              }
            }
            testID="RNE__Image__placeholder"
          >
            <View
              accessibilityLabel="loading..."
              accessibilityRole="none"
              accessible={false}
              onLayout={[Function]}
              style={
                [
                  {
                    "borderRadius": 2,
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#bdc6cf",
                    "height": 12,
                    "width": "100%",
                  },
                  undefined,
                  {
                    "height": "100%",
                    "width": "100%",
                  },
                ]
              }
              testID="RNE__Skeleton"
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgb(166, 178, 190)",
                    "height": "100%",
                    "transform": [
                      {
                        "translateX": -0,
                      },
                    ],
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={{}}
          testID="RNE__Image__children__container"
        />
      </View>
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 16,
            "fontWeight": "700",
            "marginBottom": 8,
          }
        }
      >
        Mock Movie
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
